### PR TITLE
Move result teardown to loops

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

See title.

This was added to the `LoggerConnector` first because the loops weren't finished yet.

The Loops base implementation has been changed so `teardown` is not called every `run` call as we don't want to teardown every time `run` finishes for the smaller granularity loops.

Part of https://github.com/PyTorchLightning/pytorch-lightning/issues/7938
Includes #8250 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 1]
```diff
@@ -222,7 +222,10 @@ def on_run_end(self) -> List[List[STEP_OUTPUT]]:
 
     def teardown(self) -> None:
         """Frees memory of tracked epoch outputs."""
-        self.epoch_output = None
+        self._epoch_output = None
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:225]: This was a bug, we use `_` everywhere else

File: CHANGELOG.md [hunk 2]
```diff
@@ -170,6 +170,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
     * Refactored trainer `_run_*` functions and separate evaluation loops ([#8065](https://github.com/PyTorchLightning/pytorch-lightning/pull/8065))
     * Refactored prediction loop interface; added new classes `PredictionLoop`, `PredictionEpochLoop` ([#7700](https://github.com/PyTorchLightning/pytorch-lightning/pull/7700), [#8077](https://github.com/PyTorchLightning/pytorch-lightning/pull/8077))
     * Removed `pytorch_lightning/trainer/predict_loop.py` ([#8094](https://github.com/PyTorchLightning/pytorch-lightning/pull/8094))
+    * Moved result teardown to the loops ([#8245](https://github.com/PyTorchLightning/pytorch-lightning/pull/8245))
```
Comment (tchaton) [inline CHANGELOG.md:173]: Mind making a test ?

Comment (carmocca) [reply CHANGELOG.md:173]: There are tests that fail already if the results are not moved to cpu

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:225]: thanks

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 3]
```diff
@@ -266,3 +266,7 @@ def on_evaluation_epoch_end(self) -> None:
         self.trainer.call_hook(hook_name)
         self.trainer.call_hook("on_epoch_end")
         self.trainer.logger_connector.on_epoch_end()
+
+    def teardown(self) -> None:
+        self._results.cpu()
+        self.epoch_loop.teardown()
```
Comment (awaelchli) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:272]: we are calling teardown again on the epoch_loop? It was already called in run, right?
shouldn't hurt but we may want to be aware of how many times it is getting called.

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
